# Lectura y preprocesamiento (data wrangling)

## Estado de la prestación del servicio

In [1]:
import pandas as pd

In [11]:
df = pd.read_csv("../data/raw/servicio_electrico.csv")
df.head()

,ID DEPATAMENTO,DEPARTAMENTO,ID MUNICIPIO,MUNICIPIO,ID LOCALIDAD,LOCALIDAD,AÑO SERVICIO,MES SERVICIO,ENERGÍA ACTIVA,ENERGÍA REACTIVA,POTENCIA MÁXIMA,DÍA DE DEMANDA MÁXIMA,FECHA DE DEMANDA MÁXIMA,PROMEDIO DIARIO EN HORAS
0,91,AMAZONAS,91001,LETICIA,91001000,LETICIA (LETICIA - AMAZONAS),2020,1,3.930.642,1.251.191,"7.768,76",lunes,2020 Jan 27 02:15:00 PM,24
1,91,AMAZONAS,91540,PUERTO NARIÑO,91540000,PUERTO NARIÑO (PUERTO NARIÑO - AMAZONAS),2020,1,103.897,36.304,"227,04",miércoles,2020 Jan 22 07:15:00 PM,24
2,91,AMAZONAS,91798,TARAPACÁ (ANM),91798000,TARAPACÁ (TARAPACÁ (ANM) - AMAZONAS),2020,1,22.864,9.277,"88,96",jueves,2020 Jan 30 07:30:00 PM,"10,19"
3,5,ANTIOQUIA,5873,VIGÍA DEL FUERTE,5873001,SAN ANTONIO DE PADUA (VIGÍA DEL FUERTE - ANTIO...,2020,1,5.617,1.381,"53,66",jueves,2020 Jan 23 07:45:00 PM,"4,13"
4,5,ANTIOQUIA,5873,VIGÍA DEL FUERTE,5873002,VEGÁEZ (VIGÍA DEL FUERTE - ANTIOQUIA),2020,1,2.217,539,"39,07",miércoles,2020 Jan 29 07:45:00 PM,"3,17"


In [12]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 6216 entries, 0 to 6215
Data columns (total 14 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   ID DEPATAMENTO            6216 non-null   int64
 1   DEPARTAMENTO              6216 non-null   str  
 2   ID MUNICIPIO              6216 non-null   int64
 3   MUNICIPIO                 6216 non-null   str  
 4   ID LOCALIDAD              6216 non-null   int64
 5   LOCALIDAD                 6216 non-null   str  
 6   AÑO SERVICIO              6216 non-null   int64
 7   MES SERVICIO              6216 non-null   int64
 8   ENERGÍA ACTIVA            6216 non-null   str  
 9   ENERGÍA REACTIVA          6216 non-null   str  
 10  POTENCIA MÁXIMA           6132 non-null   str  
 11  DÍA DE DEMANDA MÁXIMA     6131 non-null   str  
 12  FECHA DE DEMANDA MÁXIMA   6132 non-null   str  
 13  PROMEDIO DIARIO EN HORAS  6216 non-null   str  
dtypes: int64(5), str(9)
memory usage: 680.0 KB


In [13]:
df = df.rename(columns={
    'ID DEPATAMENTO': 'id_departamento',  # ojo: viene con typo DEPATAMENTO
    'DEPARTAMENTO': 'departamento',
    'ID MUNICIPIO': 'id_municipio',
    'MUNICIPIO': 'municipio',
    'ID LOCALIDAD': 'id_localidad',
    'LOCALIDAD': 'localidad',
    'AÑO SERVICIO': 'anio_servicio',
    'MES SERVICIO': 'mes_servicio',
    'ENERGÍA ACTIVA': 'energia_activa',
    'ENERGÍA REACTIVA': 'energia_reactiva',
    'POTENCIA MÁXIMA': 'potencia_maxima',
    'DÍA DE DEMANDA MÁXIMA': 'dia_demanda_max',
    'FECHA DE DEMANDA MÁXIMA': 'fecha_demanda_max',
    'PROMEDIO DIARIO EN HORAS': 'promedio_horas'
})

df.head()

,id_departamento,departamento,id_municipio,municipio,id_localidad,localidad,anio_servicio,mes_servicio,energia_activa,energia_reactiva,potencia_maxima,dia_demanda_max,fecha_demanda_max,promedio_horas
0,91,AMAZONAS,91001,LETICIA,91001000,LETICIA (LETICIA - AMAZONAS),2020,1,3.930.642,1.251.191,"7.768,76",lunes,2020 Jan 27 02:15:00 PM,24
1,91,AMAZONAS,91540,PUERTO NARIÑO,91540000,PUERTO NARIÑO (PUERTO NARIÑO - AMAZONAS),2020,1,103.897,36.304,"227,04",miércoles,2020 Jan 22 07:15:00 PM,24
2,91,AMAZONAS,91798,TARAPACÁ (ANM),91798000,TARAPACÁ (TARAPACÁ (ANM) - AMAZONAS),2020,1,22.864,9.277,"88,96",jueves,2020 Jan 30 07:30:00 PM,"10,19"
3,5,ANTIOQUIA,5873,VIGÍA DEL FUERTE,5873001,SAN ANTONIO DE PADUA (VIGÍA DEL FUERTE - ANTIO...,2020,1,5.617,1.381,"53,66",jueves,2020 Jan 23 07:45:00 PM,"4,13"
4,5,ANTIOQUIA,5873,VIGÍA DEL FUERTE,5873002,VEGÁEZ (VIGÍA DEL FUERTE - ANTIOQUIA),2020,1,2.217,539,"39,07",miércoles,2020 Jan 29 07:45:00 PM,"3,17"


In [14]:
# Energia y potencia: el punto es miles -> se quita; la coma es decimal -> punto
cols_miles = ['energia_activa', 'energia_reactiva', 'potencia_maxima']

for col in cols_miles:
    df[col] = df[col].astype(str).str.replace('.', '', regex=False).str.replace(',', '.', regex=False)
    df[col] = pd.to_numeric(df[col], errors='coerce')

# Horas 0-24: la coma es decimal; el punto YA es decimal ingles (2.400 = 2.4h) -> no quitarlo
df['promedio_horas'] = df['promedio_horas'].astype(str).str.replace(',', '.', regex=False)
df['promedio_horas'] = pd.to_numeric(df['promedio_horas'], errors='coerce')
df['promedio_horas'] = df['promedio_horas'].clip(upper=24)

cols_num = cols_miles + ['promedio_horas']
df[cols_num].dtypes

energia_activa        int64
energia_reactiva      int64
potencia_maxima     float64
promedio_horas      float64
dtype: object

In [15]:
df[cols_num].describe()

,energia_activa,energia_reactiva,potencia_maxima,promedio_horas
count,6.216000e+03,6.216000e+03,6.132000e+03,6216.000000
mean,3.488597e+05,9.900732e+04,1.843381e+05,11.672980
std,1.799202e+06,4.723953e+05,5.941858e+06,7.580571
min,0.000000e+00,0.000000e+00,0.000000e+00,0.000000
25%,4.041750e+03,1.482000e+03,2.852500e+01,6.150000
50%,9.216000e+03,3.862500e+03,5.997500e+01,8.617500
75%,6.308450e+04,2.247100e+04,2.297250e+02,19.032500
max,2.062621e+07,5.435213e+06,3.145708e+08,24.000000


In [16]:
df['fecha_demanda_max'] = pd.to_datetime(df['fecha_demanda_max'], format='mixed', dayfirst=True, errors='coerce')
df.isna().sum()

C:\Users\qluis\AppData\Local\Temp\ipykernel_5772\230801454.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['fecha_demanda_max'] = pd.to_datetime(df['fecha_demanda_max'], dayfirst=True, errors='coerce')


id_departamento       0
departamento          0
id_municipio          0
municipio             0
id_localidad          0
localidad             0
anio_servicio         0
mes_servicio          0
energia_activa        0
energia_reactiva      0
potencia_maxima      84
dia_demanda_max      85
fecha_demanda_max    84
promedio_horas        0
dtype: int64

In [17]:
# Texto: quitar espacios y pasar a minuscula para evitar duplicados (LETICIA vs Leticia)
cols_txt = ['departamento', 'municipio', 'localidad', 'dia_demanda_max']

for col in cols_txt:
    df[col] = df[col].astype(str).str.strip().str.lower()

df[cols_txt].head()

,departamento,municipio,localidad,dia_demanda_max
0,amazonas,leticia,leticia (leticia - amazonas),lunes
1,amazonas,puerto nariño,puerto nariño (puerto nariño - amazonas),miércoles
2,amazonas,tarapacá (anm),tarapacá (tarapacá (anm) - amazonas),jueves
3,antioquia,vigía del fuerte,san antonio de padua (vigía del fuerte - antio...,jueves
4,antioquia,vigía del fuerte,vegáez (vigía del fuerte - antioquia),miércoles


In [19]:
# IDs: el shape trae texto con cero a la izquierda ('05'); el CSV trae entero (5). Se igualan a texto.
# 1 fila trae id 0 siendo CAUCA (mpio 19809): se corrige a 19
df['id_departamento'] = df['id_departamento'].replace(0, 19)

# DANE: depto 2 digitos, mpio 5, localidad 8. zfill conserva el cero inicial.
df['id_departamento'] = df['id_departamento'].astype(str).str.zfill(2)
df['id_municipio'] = df['id_municipio'].astype(str).str.zfill(5)
df['id_localidad'] = df['id_localidad'].astype(str).str.zfill(8)

df[['id_departamento', 'id_municipio', 'id_localidad']].dtypes

id_departamento    str
id_municipio       str
id_localidad       str
dtype: object

In [18]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 6216 entries, 0 to 6215
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   id_departamento    6216 non-null   int64         
 1   departamento       6216 non-null   str           
 2   id_municipio       6216 non-null   int64         
 3   municipio          6216 non-null   str           
 4   id_localidad       6216 non-null   int64         
 5   localidad          6216 non-null   str           
 6   anio_servicio      6216 non-null   int64         
 7   mes_servicio       6216 non-null   int64         
 8   energia_activa     6216 non-null   int64         
 9   energia_reactiva   6216 non-null   int64         
 10  potencia_maxima    6132 non-null   float64       
 11  dia_demanda_max    6131 non-null   str           
 12  fecha_demanda_max  6132 non-null   datetime64[us]
 13  promedio_horas     6216 non-null   float64       
dtypes: datetime64[us](1

## Decisiones tecnicas del wrangling

**1. Lectura (`info`).** 6216 filas, 14 columnas. 5 `int64` bien; 9 `str` donde 4 debian ser numero y 1 fecha. Formato colombiano: miles con punto (`3.930.642`) y decimal con coma (`7.768,76`).

**2. Nombres (`rename`).** Minuscula `snake_case`, sin tildes ni `Ñ` (`anio_servicio`, `energia_activa`). Se corrige el typo de origen `ID DEPATAMENTO` -> `id_departamento`.

**3. Numeros.** Energia y potencia: el punto es miles -> se quita; la coma es decimal -> punto (`to_numeric`). Horas 0-24: la coma es decimal pero el punto YA es decimal ingles (`2.400` = 2.4 h) -> no quitarlo; se topa con `clip(upper=24)`. Sin esto el `describe` daba `max 2400`.

**4. Fecha (`to_datetime`).** `format='mixed', dayfirst=True` para el formato `2020 Jan 27 02:15:00 PM` sin `UserWarning`.

**5. Texto.** `strip().lower()` en departamento/municipio/localidad/dia para evitar duplicados (`LETICIA` vs `Leticia`). Mayuscula solo al unir con el shape del DANE.

**6. Nulos (`isna`).** 84 en potencia/dia/fecha son las mismas filas: Choco y Buenaventura, ene-2025, con energia y horas en 0 (sin servicio no hay pico que medir). Se exportan como `NaN`; el clustering filtra con `promedio_horas > 0`.

In [20]:
df.info()
df.to_csv('../data/processed/servicio_electrico_limpio.csv', index=False)

<class 'pandas.DataFrame'>
RangeIndex: 6216 entries, 0 to 6215
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   id_departamento    6216 non-null   str           
 1   departamento       6216 non-null   str           
 2   id_municipio       6216 non-null   str           
 3   municipio          6216 non-null   str           
 4   id_localidad       6216 non-null   str           
 5   localidad          6216 non-null   str           
 6   anio_servicio      6216 non-null   int64         
 7   mes_servicio       6216 non-null   int64         
 8   energia_activa     6216 non-null   int64         
 9   energia_reactiva   6216 non-null   int64         
 10  potencia_maxima    6132 non-null   float64       
 11  dia_demanda_max    6131 non-null   str           
 12  fecha_demanda_max  6132 non-null   datetime64[us]
 13  promedio_horas     6216 non-null   float64       
dtypes: datetime64[us](1